### PR TITLE
expression: rewrite builtin function: TIMESTAMPADD

### PR DESCRIPTION
**Pull request by spongedu**
for #4080
It's interesting that TIMESTAMPADD's return type is STRING, not DATETIME. As the following example demonstrates. My mysql version: 5.7.17 MySQL Community Server (GPL)

```
mysql> SELECT TIMESTAMPADD(WEEK,1,'2003-01-02');
Field   1:  `TIMESTAMPADD(WEEK,1,'2003-01-02')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       STRING
Collation:  utf8_general_ci (33)
Length:     87
Max_length: 10
Decimals:   31
Flags:


+-----------------------------------+  
| TIMESTAMPADD(WEEK,1,'2003-01-02') |  
+-----------------------------------+  
| 2003-01-09                        |  
+-----------------------------------+  
1 row in set (0.00 sec)
```

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -2729,37 +2729,49 @@ func (c *timestampAddFunctionClass) getFunction(args []Expression, ctx context.C
...
-func (b *builtinTimestampAddSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinTimestampAddSig) evalString(row []types.Datum) (string, bool, error) {
+	ctx := b.getCtx().GetSessionVars().StmtCtx
+	unit, isNull, err := b.args[0].EvalString(row, ctx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	v, isNull, err := b.args[1].EvalInt(row, ctx)
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
if isNull || err != nil {
    return "", isNull, errors.Trace(err)
}
```

File: expression/builtin_time.go [hunk 2]
```diff
@@ -2729,37 +2729,49 @@ func (c *timestampAddFunctionClass) getFunction(args []Expression, ctx context.C
...
-		return d, errors.Trace(err)
+		return "", isNull, errors.Trace(err)
 	}
-	unit, err := args[0].ToString()
+	if isNull {
+		return "", isNull, nil
+	}
+	arg, isNull, err := b.args[2].EvalString(row, ctx)
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 3]
```diff
@@ -2729,37 +2729,49 @@ func (c *timestampAddFunctionClass) getFunction(args []Expression, ctx context.C
...
 	}
-	if args[1].IsNull() || args[2].IsNull() {
-		return
+	if isNull {
+		return "", isNull, nil
 	}
-	v := args[1].GetInt64()
-	sc := b.ctx.GetSessionVars().StmtCtx
-	date, err := convertDatumToTime(sc, args[2])
+	date, err := types.ParseTime(arg, mysql.TypeDatetime, types.MaxFsp)
```
Comment (zz-jason) [inline expression/builtin_time.go]: maybe it's better to `bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt, tpTime)` in line 2732 and call `EvalTime` in line 2761

File: expression/builtin_time.go [hunk 4]
```diff
@@ -2784,15 +2796,14 @@ func (b *builtinTimestampAddSig) eval(row []types.Datum) (d types.Datum, err err
 	case "YEAR":
 		tb = tm1.AddDate(int(v), 0, 0)
 	default:
-		return d, errors.Trace(types.ErrInvalidTimeFormat)
+		return "", true, errors.Trace(types.ErrInvalidTimeFormat)
 	}
 	r := types.Time{Time: types.FromGoTime(tb), Type: mysql.TypeDatetime, Fsp: fsp}
 	err = r.Check()
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
if err = r.Check(); err != nil {
    ...
}
```

Comment (spongedu) [inline expression/builtin_time.go]: done

File: expression/builtin_time.go (hunk 1 above)
Comment (spongedu) [inline expression/builtin_time.go]: ok

File: expression/builtin_time.go (hunk 2 above)
Comment (spongedu) [inline expression/builtin_time.go]: ok

File: expression/builtin_time.go (hunk 3 above)
Comment (spongedu) [inline expression/builtin_time.go]: done

File: expression/builtin_time.go [hunk 5]
```diff
@@ -2729,37 +2729,42 @@ func (c *timestampAddFunctionClass) getFunction(args []Expression, ctx context.C
...
-	if args[1].IsNull() || args[2].IsNull() {
-		return
+	arg, isNull, err := b.args[2].EvalTime(row, ctx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
 	}
-	v := args[1].GetInt64()
-	sc := b.ctx.GetSessionVars().StmtCtx
-	date, err := convertDatumToTime(sc, args[2])
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: can be removed

Comment (spongedu) [inline expression/builtin_time.go]: ok

**Comment by zz-jason** [comment]
LGTM, @XuHuaiyu @winkyao @breeswish @jackysp PTAL

**Comment by jackysp** [comment]
Please resolve the conflicts.

**Comment by breeswish** [comment]
Could you add more integration tests and type infer tests? For example, tests fractional `interval` parameter and non-string `datetime_expr`. Those tests seems to be obvious for current implementation but it would be helpful to ensure that future modifications won't break it.

**Comment by breeswish** [comment]
LGTM

**Pull request merged**